src/server/engine_shard_set.cc

```diff
@@ -621,4 +692,18 @@ void EngineShardSet::TEST_EnableCacheMode() {
   RunBriefInParallel([](EngineShard* shard) { shard->db_slice().TEST_EnableCacheMode(); });
 }
 
+ShardId Shard(std::string_view v, ShardId shard_num) {
+  if (ClusterConfig::IsEnabledOrEmulated()) {
+    v = ClusterConfig::KeyTag(v);
+  }
+
+  XXH64_hash_t hash = XXH64(v.data(), v.size(), 120577240643ULL);
+
+  if (auto round_robin = g_round_robin_sharder.TryGetShardId(v, hash); round_robin.has_value()) {
+    return *round_robin;
+  }
```

Comment:
Re/ inline: we compile with lto, and also I don't know why one might think that this function should be inlined over many others. Binary size also matters, and this function is rarely called many times over (although it sometimes is).
Generally compilers are better at inlining than humans, but we can also compile with FDO if we want to optimize that, not assume blindly (imo)

Re/ other performance degradation:
1. Accessing thread locals isn't any slower than accessing global variables, and they might be in cache since we access them on every request (although that's definitely not certain). Why would accessing a single (thread-local) var with 3 members be faster than accessing 3 (thread-local) vars?
(note that these thread locals are defined and initialized in the anonymous namespace, so the compiler does not need to add locks around their initializers)

The overhead, when disabled, should be roughly to inspect the thread-local string for if it's empty. If enabled, there's also the overhead of checking the prefix-matching. Would you like me to run benchmarks for these cases? (How?)

Reply:
i think accessing one threadlocal with data members is a bit faster than accessing 3 threadlocals. You can check it via godbolt.

Reply:
I wrote a quick benchmark, and it seems to be equivalent: https://quick-bench.com/q/ofxS9x0ta4R_zuX2cdelFcDefXc